### ISSUE TITLE
Projecting velocity vectors on precomputed embeddings

### ISSUE THREAD

**Issue by tt2190** [aristoteleo/dynamo-release#256]
Hi,

Thanks for developing this great new tool.
I am wondering whether/how I can compute and project the velocity vectors onto the scanpy-based precomputed UMAP space.

I have analysed my own 10x scRNAseq data by scanpy based on the standard cellranger count outputs.
And now I have also counted spliced and unspliced transcripts with velocyto that I want to use to compute the velocity vectors to overlay on the existing UMAP.

I am thinking of something like this:
```
##############################################
# scanpy precomputation
adata = sc.read_10x_h5('my/cellranger/feature_bc_matrix.h5')
sc.pp.normalize_per_cell(adata, counts_per_cell_after=1e4)
sc.pp.log1p(adata)
sc.pp.highly_variable_genes(adata)
sc.pp.scale(adata)
sc.tl.pca(adata)
sc.pp.neighbors(adata)
sc.tl.umap(adata)

##############################################
# velocity
loom = sc.read_loom('/my/velocyto/count.loom')

# What parameters/embeddings should be transferred to "bypass" the dyn.pp.recipe_monocle step?
loom.obsm['X_pca'] = adata.obsm['X_pca']
loom.obsm['X_umap'] = adata.obsm['X_umap']

# dyn.pp.recipe_monocle(loom) # can be omitted if precomputed paramters/embeddings are added?
dyn.tl.dynamics(loom)
dyn.tl.cell_velocities(loom)
dyn.pl.streamline_plot(loom, color=['CellType'], basis='umap')
```

Simply put, my question is whether (and how) the velocity vectors can be projected on a precomputed UMAP space.

But, since ```dyn.pp.recipe_monocle``` is a huge all-in-one preprocessing function, I couldn't fully figure out what paramters/embeddings I should transfer from the precomputed anndata so that the following ```dyn.tl.dynamics``` and ```dyn.tl.cell_velocities``` work without ```dyn.pp.recipe_monocle```.

Many thanks,

**Comment by Xiaojieqiu** [comment]
hey @tt2190, thanks for using our tool. To use other embedding for velocity projection, try something like this: 
```
loom.obsm['X_umap_others'] = adata[loom.obs_names].obsm['X_umap']
dyn.tl.cell_velocities(loom, basis='umap_others', enforce=Trye)
dyn.pl.streamline_plot(loom, color=['CellType'], basis='umap_others')
```
Let me know how it goes

**Comment by tt2190** [comment]
Hi @Xiaojieqiu, thanks for your quick response. I did this and got an error below:
```
dyn.pp.recipe_monocle(loom)
dyn.tl.dynamics(loom)
dyn.tl.neighbors(loom)
loom.obsm['X_umap_others'] = adata[loom.obs_names].obsm['X_umap']
dyn.tl.cell_velocities(loom, basis='umap_others', enforce=True)
dyn.pl.streamline_plot(loom, color=['CellType'], basis='umap_others')
```

```
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
/tmp/ipykernel_117022/1720715838.py in <module>
      1 dyn.tl.neighbors(loom)
----> 2 dyn.tl.cell_velocities(loom, basis='umap_others', enforce=True)
      3 dyn.pl.streamline_plot(loom, color=['CellType'], basis='umap_others')

~/.conda/envs/velocity/lib/python3.8/site-packages/dynamo/tools/cell_velocities.py in cell_velocities(adata, ekey, vkey, X, V, X_embedding, transition_matrix, use_mnn, n_pca_components, transition_genes, min_r2, min_alpha, min_gamma, min_delta, basis, neigh_key, adj_key, add_transition_key, add_velocity_key, n_neighbors, method, neg_cells_trick, calc_rnd_vel, xy_grid_nums, correct_density, scale, sample_fraction, random_seed, enforce, preserve_len, **kernel_kwargs)
    458             )
    459         else:
--> 460             T, delta_X, X_grid, V_grid, D = kernels_from_velocyto_scvelo(
    461                 X,
    462                 X_embedding,

~/.conda/envs/velocity/lib/python3.8/site-packages/dynamo/tools/cell_velocities.py in kernels_from_velocyto_scvelo(X, X_embedding, V, indices, neg_cells_trick, xy_grid_nums, kernel, n_recurse_neighbors, max_neighs, transform, use_neg_vals, correct_density)
    965     T.eliminate_zeros()
    966 
--> 967     delta_X = projection_with_transition_matrix(n, T, X_embedding, correct_density)
    968 
    969     X_grid, V_grid, D = velocity_on_grid(

~/.conda/envs/velocity/lib/python3.8/site-packages/dynamo/tools/cell_velocities.py in projection_with_transition_matrix(n, T, X_embedding, correct_density)
    988             diff_emb = X_embedding[idx] - X_embedding[i, None]
    989             diff_emb /= norm(diff_emb, axis=1)[:, None]
--> 990             diff_emb[np.isnan(diff_emb)] = 0
    991             T_i = T[i].data
    992             delta_X[i] = T_i.dot(diff_emb)

~/.conda/envs/velocity/lib/python3.8/site-packages/anndata/_core/views.py in __setitem__(self, idx, value)
     29             )
     30             with self._update() as container:
---> 31                 container[idx] = value
     32 
     33     @contextmanager

IndexError: boolean index did not match indexed array along dimension 0; dimension is 36943 but corresponding boolean dimension is 329
```

What is the problem here?

**Comment by Xiaojieqiu** [comment]
this is an error I never see befoe. but it looks like there is a mismatch of gene dimensions during velocity projection. have you set transition_genes in `loom.var` already? and try also install dynamo by set up the latest version from github. 

If you share with me an example dataset, I can  debug it and provide more direct suggestions

**Comment by Laura19993** [comment]
Hello,

was this issue resolved? I would also like to add my own precalculated embedding (from R), but I am not sure about the workflow. I added the embedding coordinates to my adata object.
Happy for any help!

Best,
Laura
